Blog post about API documentation rollout (#411)
openSUSE/obs-landing: _posts/development/2023-06-06-api-docs-rollout.md

---
layout: post

> title: The New API Documentation Is Rolled Out!

Reviewer: ```suggestion
title: OpenAPI Specification for the Open Build Service HTTP API
```

Reviewer: @eduardoj are we using a specific OpenAPI spec version?

category: development
---

Starting today, the documentation presented through Swagger UI using the OpenAPI definition becomes the official API documentation for Open Build Service 🎉

> The team's journey to remake the API documentation has come to an end, but the most exciting journey is about to begin!

Reviewer: ```suggestion
Our journey to remake the API documentation has come to an end, but the most exciting journey is about to begin.
```

> One where any member of the OBS community can contribute to the OBS API documentation and use the OBS API easily and efficiently 👏

Reviewer: ```suggestion
Your journey to get, use and contribute to the OBS API documentation! 👏
```

{% include partials/_series-of-posts-about-api-docs.md %}

> With the rollout, we are moving away from:

Reviewer: Rollout is a concept not many of your customers will be aware of. I wouldn't mention it.

- an API documentation that wasn't very approachable for developers;

> - documentation that heavily relied on XML examples and schema files, two concepts that weren't clear enough for contributors;

Reviewer: We still heavily use both no?

- a tool ([restility](https://github.com/openSUSE/open-build-service/tree/master/docs/api/restility)) that was no longer maintained.

From now on, we all can benefit from:

- a more modern and maintained API documentation;
- documentation which is easier to understand for developers who want to make use of the OBS API in their code;

> - a more standardized way to document the API, which might increase the number of contributions.

Reviewer: ```suggestion
- a more standardized way to document the API, which hopefully increase the number of contributions.
```

> To summarize, we now provide a way to create and update the OBS API, easier to read for OBS newcomers and long-time users, and easier to document for OBS developers. What a great achievement!

Reviewer: I find self back-patting super cheesy... Let others say what kind of achievement this is. But that might just be me... :)

> Enjoy the new docu! 📖 The URL is [https://api.opensuse.org/apidocs/](https://api.opensuse.org/apidocs/). The discontinued documentation will be available in [https://build.opensuse.org/apidocs-old/](https://build.opensuse.org/apidocs-old/index) for a few more weeks.

Reviewer: Famous last words, I bet you this will be around for a very long time. Don't promise what you can't keep...

The main steps are:

- Declare the new endpoint in [OBS-v2.10.50.yaml](https://github.com/openSUSE/open-build-service/blob/master/src/api/public/apidocs/OBS-v2.10.50.yaml). Please keep the alphabetical order.
- Define the endpoint in YAML format under the [paths/](https://github.com/openSUSE/open-build-service/tree/master/src/api/public/apidocs/paths) directory.
- Make use of the already-defined [components](https://github.com/openSUSE/open-build-service/tree/master/src/api/public/apidocs/components) (parameters, responses, schemas) when possible.

You can also have a look to [closed pull requests](https://github.com/openSUSE/open-build-service/pulls?q=is%3Apr+label%3A%22Documentation+%3Abook%3A%22+is%3Aclosed) to take what others did before as an example.

Enjoy the new docu 📖 and keep the feedback and contributions coming, please. It will help the whole OBS community.

{% include partials/_how-to-give-us-feedback.md %}

> And a wholeheartedly thank you to everyone that has contributed and gave us feedback during the development of this.

Reviewer 2: ```suggestion
And a wholeheartedly thank you to everyone that has created the initial API documentation, and later on, contributed and gave us feedback during the development of this new documentation.
```

I would also mention the people who created the initial documentation, which served us as base to create the new one.

Also, finishing the sentence with "... of this." sounded weird.

Reviewer 2: By "initial API documentation" I mean "now discontinued API documentation". I will modify this suggestion in another comment.

A picture is worth a thousand words.

<figure>
  <img src="/images/posts/sprint_142_new_api_docu.gif" alt="Gif showing the usage of the new API documentation" />
  <figcaption>Usage of the new API documentation</figcaption>
</figure>

**NOTE:** The discontinued documentation will be available at [https://build.opensuse.org/apidocs-old/](https://build.opensuse.org/apidocs-old/index) for a while.

### Want To Contribute?

First of all, you need to understand what is Swagger UI and learn the OpenAPI 3.0.0 Specification. [This OpenAPI Guide](https://swagger.io/docs/specification/about/) is a good starting point.

Once you are familiar with the OpenAPI specification structure, you can start contributing to OBS documentation, which you can find under [src/api/public/apidocs](https://github.com/openSUSE/open-build-service/tree/master/src/api/public/apidocs) in our GitHub repository.

Have a look to [closed pull requests](https://github.com/openSUSE/open-build-service/pulls?q=is%3Apr+label%3A%22Documentation+%3Abook%3A%22+is%3Aclosed) to take what others did before as an example. The main steps are:

> - Declare the new endpoint in [OBS-v2.10.50.yaml](https://github.com/openSUSE/open-build-service/blob/master/src/api/public/apidocs/OBS-v2.10.50.yaml). Please keep the alphabetical order.

Reviewer 2: A contribution to the documentation will not necessarily mean adding an endpoint. I would rephrase this as:
```suggestion
- If you are documenting a new endpoint, add it in [OBS-v2.10.50.yaml](https://github.com/openSUSE/open-build-service/blob/master/src/api/public/apidocs/OBS-v2.10.50.yaml). Please keep the alphabetical order.
```

Reviewer 3: I think the correct way is `add it to ...`

Reviewer: > Please keep the alphabetical order.

I think this is too much information.

> - Define the endpoint in YAML format under the [paths/](https://github.com/openSUSE/open-build-service/tree/master/src/api/public/apidocs/paths) directory.

Reviewer 2: Also here, contributions will not always mean to define new endpoints. I also find `in YAML format` kind of redundant. I would use something like:
```suggestion
- Define or modify how endpoints behave under the [paths/](https://github.com/openSUSE/open-build-service/tree/master/src/api/public/apidocs/paths) directory.
```
